main/cache/cache.cc

```diff
@@ -21,13 +22,116 @@ unique_ptr<OwnedKeyValueStore> maybeCreateKeyValueStore(shared_ptr<::spdlog::log
                                                                       move(flavor), opts.maxCacheSizeBytes));
 }
 
+namespace {
+
+/**
+ * Returns 'true' if the given GlobalState was originally created from the current contents of
+ * kvstore (e.g., kvstore has not since been modified).
+ */
+bool kvstoreUnchangedSinceGsCreation(const core::GlobalState &gs, const unique_ptr<OwnedKeyValueStore> &kvstore) {
+    const uint8_t *maybeGsBytes = kvstore->read(core::serialize::Serializer::GLOBAL_STATE_KEY).data;
+    const bool storedUidMatches =
+        maybeGsBytes && gs.kvstoreUuid == core::serialize::Serializer::loadGlobalStateUUID(gs, maybeGsBytes);
+    const bool noPreviouslyStoredUuid = !maybeGsBytes && gs.kvstoreUuid == 0;
+    return storedUidMatches || noPreviouslyStoredUuid;
+}
+
+/**
+ * Writes the GlobalState to kvstore, but only if it was modified. Returns 'true' if a write happens.
+ */
+bool retainGlobalState(core::GlobalState &gs, const unique_ptr<OwnedKeyValueStore> &kvstore) {
+    if (kvstore && gs.wasModified() && !gs.hadCriticalError()) {
+        // Verify that no other GlobalState was written to kvstore between when we read GlobalState and wrote it
+        // into the database.
+        if (kvstoreUnchangedSinceGsCreation(gs, kvstore)) {
+            // Generate a new UUID, since this GS has changed since it was read.
+            gs.kvstoreUuid = Random::uniformU4();
+            kvstore->write(core::serialize::Serializer::GLOBAL_STATE_KEY,
+                           core::serialize::Serializer::storePayloadAndNameTable(gs));
+            return true;
+        }
+    }
+    return false;
+}
+
+bool cacheTreesAndFiles(const core::GlobalState &gs, WorkerPool &workers, absl::Span<const ast::ParsedFile> parsedFiles,
+                        const unique_ptr<OwnedKeyValueStore> &kvstore) {
+    if (kvstore == nullptr) {
+        return false;
+    }
+
+    Timer timeit(gs.tracer(), "pipeline::cacheTreesAndFiles");
```

Comment:
```suggestion
    Timer timeit(gs.tracer(), "cache::cacheTreesAndFiles");
```

Reply:
Thanks–I intentionally left this here because I didn't want to change the name of the associated metric.